### PR TITLE
ENH: xz compression in to_csv() resolves #11852

### LINKED ISSUE #11852: xz compression in to_csv()
I use compression directly in DataFrame.to_csv() to save on disk space / IO. Would be nice to have support for xz compression there.

Similar to https://github.com/pydata/pandas/issues/8654, but I think in to_csv this has more added value.

### PR DESCRIPTION
**Pull request by terfilip**
- [x] closes #11852
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.18.1.txt [hunk 1]
```diff
@@ -43,6 +43,7 @@ Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
 - ``pd.read_msgpack()`` now always gives writeable ndarrays even when compression is used (:issue:`12359`).
+- It is possible to use xz compression when writing a DataFrame to CSV (:issue:`11852`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: and reading

File: doc/source/install.rst [hunk 2]
```diff
@@ -271,6 +271,7 @@ Optional Dependencies
   `httplib2 <http://pypi.python.org/pypi/httplib2>`__
   and `google-api-python-client <http://github.com/google/google-api-python-client>`__
   : Needed for :mod:`~pandas.io.gbq`
+* `Backports.lzma <https://pypi.python.org/pypi/backports.lzma/>`__: Only for Python 2, for writing a DataFrame to an xz compressed CSV.
```
Comment (jreback) [inline doc/source/install.rst]: writing/reading

**Comment by jreback** [comment]
can do this on top of #12175 as that added compression inference and reorged the testing

File: pandas/io/common.py [hunk 3]
```diff
@@ -360,6 +360,12 @@ def _get_handle(path, mode, encoding=None, compression=None):
         elif compression == 'bz2':
             import bz2
             f = bz2.BZ2File(path, mode)
+        elif compression == 'xz':
+            try:
+                import lzma
+            except ImportError:
+                from backports import lzma
```
Comment (jreback) [inline pandas/io/common.py]: only do this if its PY2 (otherwise the ``ImportError`` should just raise)

File: pandas/parser.pyx [hunk 4]
```diff
@@ -563,6 +563,16 @@ cdef class TextReader:
                 else:
                     raise ValueError('Python 2 cannot read bz2 from open file '
                                      'handle')
+            elif self.compression == 'xz':
+                try:
+                    import lzma
+                except ImportError:
+                    from backports import lzma
```
Comment (jreback) [inline pandas/parser.pyx]: same here

**Comment by terfn** [comment]
Updated docs to include reading, using PY3 / PY2 instead of except, will rebase when #12175 gets merged.

**Comment by jreback** [comment]
ok #12175 was merged, pls rebase

**Comment by jreback** [comment]
also note #12688

**Comment by jreback** [comment]
interesting this only appears in 2 locations (and not the 3rd). any idea?

**Comment by terfn** [comment]
@jreback Rebased

File: doc/source/whatsnew/v0.18.1.txt [hunk 5]
```diff
@@ -58,6 +58,7 @@ Other Enhancements
 
 - ``pd.read_csv()`` now supports opening ZIP files that contains a single CSV, via extension inference or explict ``compression='zip'`` (:issue:`12175`)
 - ``pd.read_msgpack()`` now always gives writeable ndarrays even when compression is used (:issue:`12359`).
+- It is possible to use xz compression when writing to and/or reading from a DataFrame in CSV (:issue:`11852`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: put the right below line 59, and word similarly.

File: pandas/parser.pyx [hunk 6]
```diff
@@ -582,6 +582,16 @@ cdef class TextReader:
                 else:
                     raise ValueError('Multiple files found in compressed '
                                      'zip file %s', str(zip_names))
+            elif self.compression == 'xz':
+                if PY3:
```
Comment (jreback) [inline pandas/parser.pyx]: there is a doc-string in ``parsers.py`` that needs updating as well as docs in ``io.rst`` (I am pushing a slight fix, so rebase on master)

File: pandas/tests/frame/test_to_csv.py [hunk 7]
```diff
@@ -985,6 +985,32 @@ def test_to_csv_compression_bz2(self):
             for col in df.columns:
                 self.assertIn(col, text)
 
+    def test_to_csv_compression_xz(self):
```
Comment (jreback) [inline pandas/tests/frame/test_to_csv.py]: need to add tests in ``io/tests/test_parser.py`` in ``TestCompression``

File: pandas/tests/frame/test_to_csv.py [hunk 8]
```diff
@@ -985,6 +985,32 @@ def test_to_csv_compression_bz2(self):
...
+
+            # explicitly make sure file is xzipped
+            if compat.PY2:
+                from backports import lzma
+            else:
+                import lzma
+            f = lzma.open(filename, 'rb')
+            text = f.read().decode('utf8')
+            f.close()
+            for col in df.columns:
```
Comment (jreback) [inline pandas/tests/frame/test_to_csv.py]: use assert_frame_equal

**Comment by terfn** [comment]
@jreback For now I will also add it to `parsers.py`

**Comment by terfn** [comment]
@jreback Updated

File: doc/source/io.rst [hunk 9]
```diff
@@ -217,14 +217,14 @@ chunksize : int, default ``None``
...
   For on-the-fly decompression of on-disk data. If 'infer', then use gzip,
-  bz2 or zip if filepath_or_buffer is a string ending in '.gz', '.bz2' or
-  '.zip', respectively, and no decompression otherwise. If using 'zip',
+  bz2, zip, or xz if filepath_or_buffer is a string ending in '.gz', '.bz2',
+  '.zip', or '.xz', respectively, and no decompression otherwise. If using 'zip',
   the ZIP file must contain only one data file to be read in.
   Set to ``None`` for no decompression.
 
-  .. versionadded:: 0.18.0 support for 'zip' compression.
+  .. versionadded:: 0.18.0 support for 'zip' and 'xz' compression.
```
Comment (jreback) [inline doc/source/io.rst]: this was a typo actually, make this 0.18.1

File: pandas/io/parsers.py [hunk 10]
```diff
@@ -158,14 +158,14 @@ class ParserWarning(Warning):
...
-    bz2 or zip if filepath_or_buffer is a string ending in '.gz', '.bz2' or
-    '.zip', respectively, and no decompression otherwise. If using 'zip',
-    the ZIP file must contain only one data file to be read in.
+    bz2, zip or xz if filepath_or_buffer is a string ending in '.gz', '.bz2',
+    '.zip', or 'xz', respectively, and no decompression otherwise. If using
+    'zip', the ZIP file must contain only one data file to be read in.
     Set to None for no decompression.
 
-    .. versionadded:: 0.18.0 support for 'zip' compression.
+    .. versionadded:: 0.18.0 support for 'zip' and 'xz' compression.
```
Comment (jreback) [inline pandas/io/parsers.py]: same here, make this 0.18.1

**Comment by jreback** [comment]
minor doc fix

**Comment by terfn** [comment]
@jreback fixed typo

**Comment by jreback** [comment]
ok, lgtm. ping when green.

**Comment by terfn** [comment]
@jreback ci passes

**Comment by terfn** [comment]
hmm it was green but now looks like it's running again, either way the 'build jobs' section passes

**Comment by terfn** [comment]
@jreback green

**Comment by jreback** [comment]
@terfn thanks!

I refactored to remove some duplicate code, but looks really good!.

**Pull request closed without merging**